code/lab/dialogs/lab_ui.cpp

```diff
@@ -0,0 +1,964 @@
+					ImGui::TreeNodeEx(node_label.c_str(),
+						ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen,
+						"%s", class_def.name);
+
+					if (ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()) {
+						getLabManager()->changeDisplayedObject(LabMode::Weapon, weapon_idx);
+					}
+				}
+				weapon_idx++;
+			}
+		}
+	}
+}
+
+void LabUi::build_weapon_list() const
+{
+	//weapon display needs to be rethought
+
+	//with_TreeNode("Weapon Classes")
+	//{
+	//	build_weapon_subtype_list();
+	//}
+}
+
+void LabUi::build_background_list() const
+{
+	SCP_vector<SCP_string> missions;
+
+	cf_get_file_list(missions, CF_TYPE_MISSIONS, NOX("*.fs2"));
+
+	static SCP_map<SCP_string, SCP_vector<SCP_string>> directories;
+
+	if (directories.empty()) {
+		for (const auto& filename : missions) {
+			auto res = cf_find_file_location((filename + ".fs2").c_str(), CF_TYPE_MISSIONS);
+
+			if (res.found) {
+				auto location = get_directory_or_vp(res.full_name.c_str());
+
+				directories[location].push_back(filename);
+			}
+		}
+	}
+
+	ImGuiTreeNodeFlags node_flags = ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen;
+
+	ImGui::TreeNodeEx(LAB_MISSION_NONE_STRING, node_flags, LAB_MISSION_NONE_STRING);
+	if (ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()) {
+		getLabManager()->Renderer->useBackground(LAB_MISSION_NONE_STRING);
+	}
+
+	for (auto const& directory : directories) {
+		auto directory_name = directory.first.c_str();
+		// if the directory name is empty, this indicates loose files in the root data directory
+		// since imgui requires tree nodes to have non-empty names, we substitute a static string here
+		if (!strlen(directory_name))
+			directory_name = "Root";
+
+		with_TreeNode(directory_name)
+		{
+			for (const auto& mission_name : directory.second) {
+				ImGui::TreeNodeEx(mission_name.c_str(), node_flags, "%s", mission_name.c_str());
+
+				if (ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()) {
+					getLabManager()->Renderer->useBackground(mission_name);
+				}
+			}
+		}
+	}
+}
+
+void LabUi::build_options_menu()
+{
+	with_Menu("Options")
+	{
+		ImGui::MenuItem("Render options", nullptr, &show_render_options_dialog);
+		ImGui::MenuItem("Object selector", nullptr, &show_object_selector);
+		ImGui::MenuItem("Object options", nullptr, &show_object_options_dialog);
+		ImGui::MenuItem("Close lab", "ESC", &close_lab);
+	}
+}
+
+void LabUi::build_toolbar_entries()
+{
+	with_MainMenuBar
+	{
+		build_options_menu();
+	}
+
+	if (close_lab) {
+		getLabManager()->notify_close();
+		close_lab = false;
+	}
+}
+
+void LabUi::create_ui()
+{
+	if (show_render_options_dialog)
+		show_render_options();
+
+	if (show_object_options_dialog)
+		show_object_options();
+
+	build_toolbar_entries();
+
+	if (show_object_selector) {
+		with_Window("Select object and background")
+		{
+
+			with_CollapsingHeader("Displayed Object")
+			{
+				build_ship_list();
+
+				build_weapon_list();
+			}
+
```

Comment:
I forgot to say anything while testing but seeing it here reminded me of the simplicity of making these kinds of changes in imGui...
I think having the background and object lists in a single window is a bit of a step back. More than once I've gone through several ships and backgrounds in one session, in lists long enough that the new arrangment would be kind of a hassle. Having them side by side in seperately scrolled lists keeps from having to bounce up and down.
On the other hand I could easily take that on as my own followup PR if you don't want to make changes to the layout at this stage

Reply:
Also done.